File: packages/paso-js/src/executor.ts
import { PasoCapability, PasoDeclaration } from './types';

export interface ExecutionRequest {
  method: string;
  url: string;
  headers: Record<string, string>;
  body?: string;
}

export interface ExecutionResult {
  request: ExecutionRequest;
  status?: number;
  statusText?: string;
  body: string;
  durationMs: number;
  error?: string;
  retryAfterSeconds?: number;
}

export interface StructuredError {
  error: true;
  type:
    | 'auth_failed'
    | 'forbidden'
    | 'not_found'
    | 'rate_limited'
    | 'server_error'
    | 'network_error'
    | 'client_error';
  status: number | null;
  message: string;
  hint: string;
  retry_after_seconds?: number;
  request_url: string;
}

/**
 * Build the HTTP request for a capability without executing it.
 * Note: does not validate required inputs — callers (CLI, MCP) must validate before calling.
 */
export function buildRequest(
  cap: PasoCapability,
  args: Record<string, unknown>,
  decl: PasoDeclaration,
  authToken?: string,
): ExecutionRequest {
  let path = cap.path;
  const queryParams: Record<string, string> = {};
  const bodyParams: Record<string, unknown> = {};

  if (cap.inputs) {
    for (const [name, input] of Object.entries(cap.inputs)) {
      const value = args[name];
      if (value === undefined) continue;

      const location =
        input.in || (['POST', 'PUT', 'PATCH'].includes(cap.method) ? 'body' : 'query');

      switch (location) {
        case 'path':
          path = path.replace(`{${name}}`, encodeURIComponent(String(value)));
          break;
        case 'query':
          queryParams[name] = String(value);
          break;
        case 'header':
          // Handled in the header loop below
          break;
        case 'body':
          bodyParams[name] = value;
          break;
        default:
          throw new Error(
            `Unknown input location "${location}" for parameter "${name}". Expected one of: path, query, body, header.`,
          );
      }
    }
  }

  let baseUrl = decl.service.base_url;
  while (baseUrl.endsWith('/')) baseUrl = baseUrl.slice(0, -1);
  const fullPath = path.startsWith('/') ? path : `/${path}`;
  const url = new URL(`${baseUrl}${fullPath}`);
  for (const [k, v] of Object.entries(queryParams)) {
    url.searchParams.set(k, v);
  }

  const hasBody = ['POST', 'PUT', 'PATCH'].includes(cap.method);
  const headers: Record<string, string> = {
    Accept: 'application/json',
    ...(hasBody ? { 'Content-Type': 'application/json' } : {}),
  };

  if (decl.service.auth) {
    if (decl.service.auth.type === 'none') {
      // Skip auth — notice is logged once at command startup, not per-request
    } else {
      const token = authToken;
      if (token) {
        const authType = decl.service.auth.type;
        const authHeader = decl.service.auth.header || 'Authorization';
        switch (authType) {
          case 'bearer':
          case 'oauth2': {
            const prefix = decl.service.auth.prefix ?? 'Bearer';
            headers[authHeader] = prefix ? `${prefix} ${token}` : token;
            break;
          }
          case 'api_key':
            headers[authHeader] = token;
            break;
          default:
            process.stderr.write(
              `Warning: unknown auth.type "${authType}" — sending token as-is in ${authHeader}\n`,
            );
            headers[authHeader] = token;
        }
      }
    }
  }

  // Header-type inputs
  if (cap.inputs) {
    for (const [name, input] of Object.entries(cap.inputs)) {
      if (input.in === 'header' && args[name] !== undefined) {
        headers[name] = String(args[name]).replace(/[\r\n]/g, '');
      }
    }
  }

  const req: ExecutionRequest = {
    method: cap.method,
    url: url.toString(),
    headers,
  };

  if (['POST', 'PUT', 'PATCH'].includes(cap.method) && Object.keys(bodyParams).length > 0) {
    req.body = JSON.stringify(bodyParams);
  }

  return req;
}

/**
 * Execute an HTTP request and return the result.
 */
export async function executeRequest(
  req: ExecutionRequest,
  options?: { timeout?: number },
): Promise<ExecutionResult> {
  const start = Date.now();
  const timeout = options?.timeout ?? 30000;

  try {
    const fetchOptions: RequestInit = {
      method: req.method,
      headers: req.headers,
      signal: AbortSignal.timeout(timeout),
    };

    if (req.body) {
      fetchOptions.body = req.body;
    }

    const MAX_RESPONSE_SIZE = 10 * 1024 * 1024; // 10MB
    const response = await fetch(req.url, fetchOptions);

    // Check content-length first (fast path)
    const contentLength = response.headers.get('content-length');
    if (contentLength && parseInt(contentLength, 10) > MAX_RESPONSE_SIZE) {
      throw new Error(`Response too large (${contentLength} bytes, max ${MAX_RESPONSE_SIZE})`);
    }

    // Read body with size guard — also catches chunked responses without content-length
    let text: string;
    if (response.body) {
      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      const chunks: string[] = [];
      let totalBytes = 0;

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        totalBytes += value.byteLength;
        if (totalBytes > MAX_RESPONSE_SIZE) {
          reader.cancel();
          throw new Error(
            `Response too large (>${MAX_RESPONSE_SIZE} bytes streamed, max ${MAX_RESPONSE_SIZE})`,
          );
        }
        chunks.push(decoder.decode(value, { stream: true }));
      }
      chunks.push(decoder.decode()); // flush
      text = chunks.join('');
    } else {
      text = await response.text();
    }

    const durationMs = Date.now() - start;

    let body: string;
    try {
      const json = JSON.parse(text);
      body = JSON.stringify(json, null, 2);
    } catch {
      body = text;
    }

    // Capture Retry-After header for 429 responses
    let retryAfterSeconds: number | undefined;
    if (response.status === 429) {
      const retryAfter = response.headers.get('retry-after');
      if (retryAfter) {
        const parsed = parseInt(retryAfter, 10);
        retryAfterSeconds = isNaN(parsed) ? undefined : parsed;
      }
    }

    return {
      request: req,
      status: response.status,
      statusText: response.statusText,
      body,
      durationMs,
      retryAfterSeconds,
    };
  } catch (error) {
    return {
      request: req,
      body: '',
      durationMs: Date.now() - start,
      error: error instanceof Error ? error.message : String(error),
    };
  }
}

/**
 * Format a friendly error message for common HTTP status codes.
 */
export function formatError(
  result: ExecutionResult,
  decl: PasoDeclaration,
  authToken?: string,
): string {
  if (result.error) {
    return `Request failed: ${result.error}`;
  }

  if (!result.status) return 'Unknown error';

  if (result.status === 401) {
    const authType = decl.service.auth?.type || 'unknown';
    const hasToken = !!authToken;
    let msg = `Error 401: Authentication failed.`;
    if (!hasToken) {
      msg += `\n  → USEPASO_AUTH_TOKEN is not set. Set it with: export USEPASO_AUTH_TOKEN=your-token`;
    } else {
      msg += `\n  → USEPASO_AUTH_TOKEN is set but was rejected by the API.`;
      msg += `\n  → Auth type: ${authType}. Check that your token is valid and has the required scopes.`;
    }
    return msg;
  }

  if (result.status === 403) {
    return `Error 403: Forbidden. Your token does not have permission for this action.\n  → Check the required scopes/permissions for this endpoint.`;
  }

  if (result.status === 404) {
    return `Error 404: Not found.\n  → Check that base_url and path are correct in your usepaso.yaml.\n  → URL was: ${result.request.url}`;
  }

  if (result.status === 429) {
    const retryHint = result.retryAfterSeconds
      ? `Retry after ${result.retryAfterSeconds} seconds.`
      : 'Wait and try again, or check your rate limit constraints.';
    return `Error 429: Rate limited. ${retryHint}`;
  }

  if (result.status >= 500) {
    return `Error ${result.status}: Server error from the API.\n  → This is likely a problem on the API side, not with usepaso.`;
  }

  return `Error ${result.status} ${result.statusText || ''}: ${result.body}`;
}

/**
 * Format a structured error for MCP tool responses.
 * Returns a JSON-serializable object that agents can parse programmatically.
 */
export function formatStructuredError(
  result: ExecutionResult,
  decl: PasoDeclaration,
  authToken?: string,
): StructuredError {
  const url = result.request.url;

  if (result.error) {
    return {
      error: true,
      type: 'network_error',
      status: null,
      message: result.error,
      hint: 'Check your network connection and the service base_url.',
      request_url: url,
    };
  }

  const status = result.status || 0;

  if (status === 401) {
    const hasToken = !!authToken;
    return {
      error: true,
      type: 'auth_failed',
      status,
      message: 'Authentication failed.',
      hint: hasToken
        ? 'USEPASO_AUTH_TOKEN was rejected. Check that it is valid and has the required scopes.'
        : 'USEPASO_AUTH_TOKEN is not set.',
      request_url: url,
    };
  }

  if (status === 403) {
    return {
      error: true,
      type: 'forbidden',
      status,
      message: 'Forbidden. Your token does not have permission for this action.',
      hint: 'Check the required scopes/permissions for this endpoint.',
      request_url: url,
    };
  }

  if (status === 404) {
    return {
      error: true,
      type: 'not_found',
      status,
      message: 'Not found.',
      hint: 'Check that base_url and path are correct in your usepaso.yaml.',
      request_url: url,
    };
  }

  if (status === 429) {
    const se: StructuredError = {
      error: true,
      type: 'rate_limited',
      status,
      message: 'Rate limited.',
      hint: 'Wait and retry.',
      request_url: url,
    };
    if (result.retryAfterSeconds) {
      se.retry_after_seconds = result.retryAfterSeconds;
      se.hint = `Retry after ${result.retryAfterSeconds} seconds.`;
    }
    return se;
  }

  if (status >= 500) {
    return {
      error: true,
      type: 'server_error',
      status,
      message: `Server error (${status}).`,
      hint: 'This is likely a problem on the API side, not with usepaso.',
      request_url: url,
    };
  }

  return {
    error: true,
    type: 'client_error',
    status,
    message: `Error ${status}.`,
    hint: result.body || 'Unexpected error.',
    request_url: url,
  };
}

File: packages/paso-py/paso/executor.py
"""
Shared HTTP execution module used by both the MCP server and the test CLI command.
"""

import asyncio
import json
import os
import time
from typing import Any, Optional
from urllib.parse import quote, urlencode

import httpx

from paso.types import PasoCapability, PasoDeclaration


def build_request(cap: PasoCapability, args: dict, decl: PasoDeclaration) -> dict:
    """
    Build an HTTP request dict from a capability, arguments, and declaration.

    Args:
        cap: PasoCapability with method, path, inputs, etc.
        args: Dictionary of argument values
        decl: PasoDeclaration with service config and auth details

    Returns:
        Dict with keys: method, url, headers, body (optional)
    """
    method = cap.method.upper()
    path = cap.path

    # Substitute path parameters
    path_params = {}
    query_params = {}
    body_params = {}
    header_params = {}

    if cap.inputs:
        for param_name, input_spec in cap.inputs.items():
            if param_name not in args:
                continue

            param_value = args[param_name]
            in_type = getattr(input_spec, "in_", None)

            # Determine location if not explicitly set
            if in_type is None:
                if method in ("GET", "DELETE"):
                    in_type = "query"
                elif method in ("POST", "PUT", "PATCH"):
                    in_type = "body"

            if in_type == "path":
                path_params[param_name] = param_value
            elif in_type == "query":
                query_params[param_name] = param_value
            elif in_type == "body":
                body_params[param_name] = param_value
            elif in_type == "header":
                header_params[param_name] = param_value
            else:
                raise ValueError(
                    f'Unknown input location "{in_type}" for parameter "{param_name}". '
                    f'Expected one of: path, query, body, header.'
                )

    # Substitute path parameters into path
    for param_name, param_value in path_params.items():
        path = path.replace(f"{{{param_name}}}", quote(str(param_value), safe=''))

    # Build URL
    base_url = decl.service.base_url.rstrip("/")
    path = path.lstrip("/")
    url = f"{base_url}/{path}"

    # Append query parameters
    if query_params:
        query_string = urlencode(query_params)
        url = f"{url}?{query_string}"

    # Build headers
    headers = {
        "Content-Type": "application/json",
        "Accept": "application/json",
    }

    # Add authentication header
    auth_token = os.environ.get("USEPASO_AUTH_TOKEN")
    if decl.service.auth:
        if decl.service.auth.type == "none":
            pass  # Skip auth — notice is logged once at command startup, not per-request
        elif auth_token:
            auth = decl.service.auth
            header_name = auth.header or "Authorization"
            if auth.type in ("bearer", "oauth2"):
                prefix = auth.prefix if auth.prefix is not None else "Bearer"
                headers[header_name] = f"{prefix} {auth_token}" if prefix else auth_token
            elif auth.type == "api_key":
                headers[header_name] = auth_token
            else:
                import sys
                print(
                    f'Warning: unknown auth.type "{auth.type}" — sending token as-is in {header_name}',
                    file=sys.stderr,
                )
                headers[header_name] = auth_token

    # Add header parameters (strip newlines to prevent header injection)
    for k, v in header_params.items():
        headers[k] = str(v).replace('\r', '').replace('\n', '')

    # Build request dict
    req = {
        "method": method,
        "url": url,
        "headers": headers,
    }

    # Add body if needed
    if method in ("POST", "PUT", "PATCH") and body_params:
        req["body"] = json.dumps(body_params)

    return req


async def execute_request(req: dict) -> dict:
    """
    Execute an HTTP request asynchronously.

    Args:
        req: Request dict from build_request with method, url, headers, and optional body

    Returns:
        Dict with: request, status, status_text, body, duration_ms, error
    """
    start_time = time.time()
    result = {
        "request": req,
        "status": None,
        "status_text": "",
        "body": "",
        "duration_ms": 0,
        "error": None,
    }

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.request(
                method=req["method"],
                url=req["url"],
                headers=req["headers"],
                content=req.get("body"),
            )

            result["status"] = response.status_code
            result["status_text"] = response.reason_phrase or ""

            # Try to pretty-print JSON responses
            try:
                body_json = response.json()
                result["body"] = json.dumps(body_json, indent=2)
            except (json.JSONDecodeError, ValueError):
                result["body"] = response.text

    except httpx.RequestError as e:
        result["error"] = str(e)
    except Exception as e:
        result["error"] = str(e)

    end_time = time.time()
    result["duration_ms"] = int((end_time - start_time) * 1000)

    return result


def format_error(result: dict, decl: PasoDeclaration) -> str:
    """
    Format a friendly error message from a request result.

    Args:
        result: Result dict from execute_request
        decl: PasoDeclaration with service config

    Returns:
        Formatted error message string
    """
    auth_token = os.environ.get("USEPASO_AUTH_TOKEN")

    # Handle connection errors
    if result["error"]:
        return f"Request failed: {result['error']}"

    status = result["status"]

    if status == 401:
        hint = ""
        if not auth_token:
            hint = "\nHint: USEPASO_AUTH_TOKEN is not set. Please set it and try again."
        else:
            hint = "\nHint: Your USEPASO_AUTH_TOKEN may be invalid or expired."
        return f"Error 401: Authentication failed.{hint}"

    elif status == 403:
        return "Error 403: Forbidden.\nHint: Your token may not have the required scopes for this endpoint."

    elif status == 404:
        url = result["request"]["url"]
        hint = f"\nHint: Check your base_url ({decl.service.base_url}) and path."
        return f"Error 404: Not found.\nURL: {url}{hint}"

    elif status == 429:
        return "Error 429: Rate limited.\nPlease wait before retrying."

    elif status and status >= 500:
        return f"Error {status}: Server error from the API."

    return f"Error {status}: {result['status_text']}"
